[Diagnostics] A collection of diagnostic fixes/improvements

xedin opened the pull request:
This fixes deal with situations that previously caused a fallback diagnostic to be produced.

- Detect when generic argument mismatch applies to argument and produce a tailed fix
- Assign missing member eagerly when there is no context
- Diagnose existential mismatch in a literal collection element position
- Use argument mismatch fix for `@autoclosure` result positions
- Don't attempt to force unwrap invalid optional chaining in the argument position

Resolves: rdar://82971941
Resolves: rdar://152021264
Resolves: https://github.com/swiftlang/swift/issues/81770
Resolves: rdar://103045274
Resolves: rdar://110527062
Resolves: rdar://126080504

<!--
If this pull request is targeting a release branch, please fill out the
following form:
https://github.com/swiftlang/.github/blob/main/PULL_REQUEST_TEMPLATE/release.md?plain=1

Otherwise, replace this comment with a description of your changes and
rationale. Provide links to external references/discussions if appropriate.
If this pull request resolves any GitHub issues, link them like so:

  Resolves <link to issue>, resolves <link to another issue>.

For more information about linking a pull request to an issue, see:
https://docs.github.com/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue
-->

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/main/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[Diagnostics] Don't attempt to force unwrap invalid optional chaining in the argument position

If the argument has an extra `?` or `!`, let's not attempt to force
optional (because it's use is already invalid) and re-introduce
the constraint with unwrapped type instead. This would help to diagnose
the invalid chaining as well as any argument to parameter mismatches.

Resolves: rdar://126080504

xedin committed:
[Diagnostics] Use argument mismatch fix for `@autoclosure` result positions

`@autoclosure` is associated with a parameter, we use argument mismatch fix
to diagnose missing explicit calls as well as any mismatches in that position.

Resolves: rdar://110527062

xedin committed:
[Diagnostics] Diagnose existential mismatch in a literal collection element position

If key or value of a literal collection expression doesn't conform
to protocol(s) expected by the contextual existential type, let's
diagnose that via a tailed collection mismatch fix instead of a
generic conformance one.

Resolves: rdar://103045274

xedin committed:
[Diagnostics] Assign missing member eagerly when there is no context

Without contextual information it won't be possible to bind a missing
member to a concrete type later, so let's bind them eagerly and propagate
placeholders outward.

Resolves: rdar://152021264
Resolves: https://github.com/swiftlang/swift/issues/81770

xedin committed:
[CSSimplify] Detect when generic argument mismatch applies to argument and produce a tailed fix

The problem detection logic currently expects `generic argument #<N>`
location to always be associated with two generic types, but that
is not always the case, this locator element is sometimes used for
i.e. optional object types and pointer `Pointee` type when types
appear in argument positions. This needs to be handled specifically.

Resolves: rdar://82971941

xedin commented:
@swift-ci please test

hamishknight commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -6884,9 +6884,28 @@ bool ConstraintSystem::repairFailures(
+        if (auto overload = findSelectedOverloadFor(getCalleeLocator(argLoc))) {
+          auto *overloadTy =
+              simplifyType(overload->boundType)->castTo<FunctionType>();
+          auto *argList = getAsExpr(anchor)->getArgs();
+          ASSERT(argList);
```
This could potentially be null for e.g a keypath subscript, I would extract the argument list using `getArgumentList` instead

xedin replied:
Good point!

xedin replied:
✅

xedin commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -5883,6 +5883,24 @@ bool ConstraintSystem::repairFailures(
       if (hasConversionOrRestriction(ConversionRestrictionKind::Existential))
         break;
 
+      if (auto *typeVar =
+              lhs->getOptionalObjectType()->getAs<TypeVariableType>()) {
+        auto *argLoc = typeVar->getImpl().getLocator();
+        if (argLoc->directlyAt<OptionalEvaluationExpr>()) {
+          auto OEE = castToExpr<OptionalEvaluationExpr>(argLoc->getAnchor());
+          // If the optional chain in the argument position is invalid
+          // let's unwrap optional and re-introduce the constraint to
+          // be solved later once both sides are sufficiently resolved,
+          // this would allow to diagnose not only the invalid unwrap
+          // but an invalid conversion (if any) as well.
+          if (hasFixFor(getConstraintLocator(OEE->getSubExpr()),
+                        FixKind::RemoveUnwrap)) {
+            addConstraint(matchKind, typeVar, rhs, loc);
+            return true;
+          }
+        }
+      }
+
```
Yeah, I knew about that but I didn't want to mess too much with this block, I think the whole `if` block here could be removed and handled by `repairViaOptionalUnwrap` farther down. Let me try to see what it's going to take to do that...

xedin replied:
I did some refactoring in `ApplyArgToParam` and move some logic that belongs to `LValueConversion` out, which means that we no longer need to fallthrough.

hamishknight replied:
Nice!

xedin committed:
[CSSimplify] Move `inout` failures handing to `LValueConversion` block

This is cleanup that allows us to avoid a fallthrough to `ApplyArgToParam`
handling for `inout` failures.

xedin committed:
[CSSimplify] Remove ad-hoc optional mismatch handling for arguments

Since we have `repairViaOptionalUnwrap`, let's use that to fix
argument/parameter mismatches instead of doing ad-hoc matching.

xedin commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -11438,6 +11438,13 @@ ConstraintSystem::SolutionKind ConstraintSystem::simplifyMemberConstraint(
             // are simplified.
             addTypeVariableConstraintsToWorkList(memberTypeVar);
+          } else if (locator->getAnchor().is<Expr *>() &&
+                     !getParentExpr(getAsExpr(locator->getAnchor()))) {
```
Fixed!

xedin commented:
@swift-ci please test

hamishknight commented on include/swift/Sema/ConstraintSystem.h:
```diff
@@ -3484,6 +3484,15 @@ class ConstraintSystem {
 
+  Expr *getSemanticsProvidingParentExpr(Expr *expr) {
+    while (auto *parent = getParentExpr(expr)) {
+      if (!(isa<IdentityExpr>(parent) || isa<TryExpr>(parent)))
```
Worth doing `parent->getSemanticsProvidingExpr() == parent` to avoid the duplication?

xedin commented:
@swift-ci please smoke test
